### ISSUE TITLE
Roslyn should not change nullability annotation of variables passed to `in` arguments

### ISSUE THREAD

**Issue by TessenR** [dotnet/roslyn#40588]
**Version Used**: 

```
Branch master (23 Dec 2019)
Latest commit bc98ebc by dotnet-maestro[bot]:
[master] Update dependencies from dotnet/arcade (#40557)

* Update dependencies from https://github.com/dotnet/arcade build 20191221.1

- Microsoft.DotNet.Arcade.Sdk - 5.0.0-beta.19621.1

* Update dependencies from https://github.com/dotnet/arcade build 20191222.1

- Microsoft.DotNet.Arcade.Sdk - 5.0.0-beta.19622.1
```

**Steps to Reproduce**:

```
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  void M1(C c)
  {
    c.ToString();
    Test(in c);
    c.ToString();
  }
    
  void Test([MaybeNull] in C c) { }
}
```

**Expected Behavior**:
No warnings. `c` is a variable of non-nullable type and there's no write to this variable since values of `in` parameters cannot be changed.

**Actual Behavior**:
The second `c.ToString()` reports incorrect warning `CS8602: Dereference of a possibly null reference.`

**Notes**
It seems that `[MaybeNull]` annotation forces Roslyn to think that the variable can change and become nullable even though it's impossible. `[MaybeNull]` annotation in fact shouldn't affect the analysis of this program at all since it is shouldn't be applicable in this scenario. If you remove the annotation there will be no warnings regardless of whether the parameter is declared as nullable or not i.e. there are no warnings in the following code:

```cs
#nullable enable
class C
{
  void M1(C c)
  {
    c.ToString();
    Test(in c);
    c.ToString();
  }
    
  void Test(in C? c) { }
}
```

It's also worth noting that in this case it's impossible to tell Roslyn that the variable cannot be assigned with `null` with the suppress nullable warning expression even though it works for other types of byref arguments i.e. the following code only reports a warning in `M3`:

```cs
#nullable enable
using System.Diagnostics.CodeAnalysis;

class C
{
  void M1(C c)
  {
    Test1(out c!); // ! tells Roslyn that null wasn't assigned
    c.ToString();
  }
    
  void M2(C c)
  {
    Test2(ref c!); // ! tells Roslyn that null wasn't assigned
    c.ToString();
  }
    
  void M3(C c)
  {
    Test3(in c!); // Roslyn just ignores !
    c.ToString(); // the only warning in the program, incorrect
  }
    
  void Test1([MaybeNull] out C? c) { c = new C(); }
  void Test2([MaybeNull] ref C? c) { }
  void Test3([MaybeNull] in C? c) { }
}
```

**Comment by sharwell** [comment]
This attribute can be used to document assertions within code flow, such as `Assert.MightBeNull(obj)`. The dotnet/machinelearning repository uses this pattern in many places.

**Comment by TessenR** [comment]
Oh thanks, I forgot about this. Guess it can be closed then
